### ISSUE TITLE
Programming question about function call for 'fwd' and 'predict' in trainer.py

### ISSUE THREAD

**Issue by User-LiL** [facebookresearch/XLM#160]
Hi, I am graduate major in machine translation.
When I read the function 'clm_step' - 'mlm_step' - in trainer.py.
I don't understand why 'model' can call the function 'fwd' and 'predict'.

657:   def clm_step(self, lang1, lang2, lambda_coeff):
.
.
683:       # forward / loss
683:       tensor = model('fwd', x=x, lengths=lengths, langs=langs, causal=True)
683:       _, loss = model('predict', tensor=tensor, pred_mask=pred_mask, y=y, get_scores=False)
.
.

Thanks a lot.

**Comment by glample** [comment]
Hi,

`fwd` takes a sequence of word indices as input (i.e. input sentences) and outputs hidden states for each input word. It basically applies the transformer.
`predict` projects the transformer output (given by `fwd`) and generates a score for each word in the vocabulary.

**Comment by User-LiL** [comment]
I know the 'model' is an object of class TransformerModel, and I guess the 'fwd' and 'predict' are usual functions in class TransformerModel.

I don't know why the 'model' could call the 'fwd' and 'predict' functions in this way.
Usually, I know the basic way for call a function: "object.method(parameters)...".

Thanks a lot.

**Comment by glample** [comment]
I'm not sure I understand your question. Yes, `fwd` and `predict` are usual functions. But the Distributed wrapper does not let you call a model custom functions, only the `__forward__` one. So, when we call the model, we call it with `__forward__` by providing `fwd` or `predict` and then the model internally calls one of these two functions.
